cranelift/filetests/src/function_runner.rs

```diff
@@ -1,96 +1,389 @@
+/// function through the use of a [Trampoline].
+///
+/// ```
+/// use cranelift_filetests::SingleFunctionCompiler;
+/// use cranelift_reader::{parse_functions, DataValue};
+///
+/// let code = "test run \n function %add(i32, i32) -> i32 {  block0(v0:i32, v1:i32):  v2 = iadd v0, v1  return v2 }".into();
+/// let func = parse_functions(code).unwrap().into_iter().nth(0).unwrap();
+/// let mut compiler = SingleFunctionCompiler::with_default_host_isa();
+/// let compiled_func = compiler.compile(func).unwrap();
+///
+/// let returned = compiled_func.call(&vec![DataValue::I32(2), DataValue::I32(40)]);
+/// assert_eq!(vec![DataValue::I32(42)], returned);
+/// ```
+pub struct CompiledFunction<'a> {
+    page: Mmap,
+    signature: Signature,
+    trampoline: &'a Trampoline,
+}
+
+impl<'a> CompiledFunction<'a> {
+    /// Build a new [CompiledFunction].
+    pub fn new(page: Mmap, signature: Signature, trampoline: &'a Trampoline) -> Self {
+        Self {
+            page,
+            signature,
+            trampoline,
+        }
+    }
+
+    /// Return a pointer to the compiled code.
+    pub fn as_ptr(&self) -> *const u8 {
+        self.page.as_ptr()
+    }
+
+    /// Call the [CompiledFunction], passing in [DataValue]s using a compiled [Trampoline].
+    pub fn call(&self, arguments: &[DataValue]) -> Vec<DataValue> {
+        let mut values = UnboxedValues::make_arguments(arguments, &self.signature);
+        let arguments_address = values.as_mut_ptr();
+        let function_address = self.as_ptr();
+
+        let callable_trampoline: fn(*const u8, *mut u128) -> () =
+            unsafe { mem::transmute(self.trampoline.as_ptr()) };
+        callable_trampoline(function_address, arguments_address);
+
+        values.collect_returns(&self.signature)
+    }
+}
+
+/// A container for laying out the [ValueData]s in memory in a way that the [Trampoline] can
+/// understand.
+struct UnboxedValues(Vec<u128>);
+
+impl UnboxedValues {
+    /// The size in bytes of each slot location in the allocated [DataValue]s. Though [DataValue]s
+    /// could be smaller than 16 bytes (e.g. `I16`), this simplifies the creation of the [DataValue]
+    /// array and could be used to align the slots to the largest used [DataValue] (i.e. 128-bit
+    /// vectors).
+    const SLOT_SIZE: usize = 16;
+
+    /// Build the arguments vector for passing the [DataValue]s into the [Trampoline]. The size of
+    /// `u128` used here must match [Trampoline::SLOT_SIZE].
+    pub fn make_arguments(arguments: &[DataValue], signature: &ir::Signature) -> Self {
+        assert_eq!(arguments.len(), signature.params.len());
+        let mut values_vec = vec![0; max(signature.params.len(), signature.returns.len())];
+
+        // Store the argument values into `values_vec`.
+        for ((arg, slot), param) in arguments.iter().zip(&mut values_vec).zip(&signature.params) {
+            if arg.ty() != param.value_type {
+                panic!("argument type mismatch");
```

Comment:
This will manifest as a panic when running a test, right? Is there a way to propagate this as an error to the test runner, with the right file location, instead?